From 2c0236685072f90eabf62130e5b0491080800fff Mon Sep 17 00:00:00 2001
From: Jammy Huang <jammy_huang@aspeedtech.com>
Date: Thu, 12 Jun 2025 10:13:41 +0800
Subject: [PATCH 4/4] Improve video quality

With current default quality, mosquito effect could be observed on
text. To fix this issue, set default quality to 4

Upstream-Status: Inappropriate [ASPEED specific]
Signed-off-by: Jammy Huang <jammy_huang@aspeedtech.com>
---
 ikvm_server.cpp |  2 +-
 ikvm_video.cpp  | 11 ++++++++++-
 ikvm_video.hpp  | 11 +++++++++++
 3 files changed, 22 insertions(+), 2 deletions(-)

diff --git a/ikvm_server.cpp b/ikvm_server.cpp
index 9e39f16..ef29a38 100644
--- a/ikvm_server.cpp
+++ b/ikvm_server.cpp
@@ -222,7 +222,7 @@ void Server::sendFrame()
                 cl->updateBuf[cl->ublen++] = (char)(frameCount >> 8 & 0xff);
                 cl->updateBuf[cl->ublen++] = (char)(frameCount & 0xff);
                 cl->updateBuf[cl->ublen++] = (char)(video.getSubsampling());
-                cl->updateBuf[cl->ublen++] = (char)(0x00); // the value of quality if changed
+                cl->updateBuf[cl->ublen++] = (char)(video.getQuality());
 
                 rfbSendCompressedDataTight(cl, data, video.getFrameSize(i));
                 if (cl->enableLastRectEncoding)
diff --git a/ikvm_video.cpp b/ikvm_video.cpp
index 44b7106..01f2d96 100644
--- a/ikvm_video.cpp
+++ b/ikvm_video.cpp
@@ -32,7 +32,7 @@ using namespace sdbusplus::xyz::openbmc_project::Common::Device::Error;
 
 Video::Video(const std::string& p, Input& input, int fr, int sub, int fmt) :
     resizeAfterOpen(false), timingsError(false), fd(-1), frameRate(fr),
-    height(600), width(800), subSampling(sub), input(input),
+    height(600), width(800), subSampling(sub), quality(4), input(input),
     path(p), pixelformat(V4L2_PIX_FMT_JPEG), pixelformatflags(0)
 {
     switch (fmt)
@@ -566,6 +566,15 @@ void Video::start()
                      strerror(errno));
     }
 
+    ctrl.id = V4L2_CID_JPEG_COMPRESSION_QUALITY;
+    ctrl.value = quality;
+    rc = ioctl(fd, VIDIOC_S_CTRL, &ctrl);
+    if (rc < 0)
+    {
+        lg2::warning("Failed to set video jpeg quality {ERROR}", "ERROR",
+                     strerror(errno));
+    }
+
     ctrl.id = V4L2_CID_JPEG_CHROMA_SUBSAMPLING;
     ctrl.value = subSampling ? V4L2_JPEG_CHROMA_SUBSAMPLING_420
                              : V4L2_JPEG_CHROMA_SUBSAMPLING_444;
diff --git a/ikvm_video.hpp b/ikvm_video.hpp
index e7d6594..2cd195b 100644
--- a/ikvm_video.hpp
+++ b/ikvm_video.hpp
@@ -166,6 +166,15 @@ class Video
     {
         subSampling = _sub;
     }
+    /*
+     * @brief Gets the quality of the jpeg compression result
+     *
+     * @return Value of the jpeg quality
+     */
+    inline int getQuality() const
+    {
+        return quality;
+    }
 
     /*
      * @brief Gets the bounding-box of the partial-jpeg
@@ -228,6 +237,8 @@ class Video
     size_t width;
     /* @brief jpeg's subsampling, 1:420/0:444 */
     int subSampling;
+    /* @brief jpeg's quality */
+    int quality;
     /* @brief Reference to the Input object */
     Input& input;
     /* @brief Path to the V4L2 video device */
-- 
2.43.0

